File: tests/testthat/test-adaptive_alpha.R
test_that("calibrate_alpha returns one row per grid point and errors without permuted arm", {
  d <- sim_cell_type(40, n_genes = 40, seed = 8)
  out <- run_downsampling(d$counts, d$group, "wilcoxon",
                           grid_args = list(n_min = 10, n_points = 3),
                           k = 2, k_permute = 3, seed = 12)
  ac <- calibrate_alpha(out, target_fpr = 0.05)
  expect_equal(nrow(ac), length(unique(out$downsampled$n)))
  expect_true(all(ac$alpha >= 0 & ac$alpha <= 1))

  out_noperm <- run_downsampling(d$counts, d$group, "wilcoxon",
                                  grid_args = list(n_min = 10, n_points = 3),
                                  k = 2, k_permute = 0, seed = 13)
  expect_error(calibrate_alpha(out_noperm), "no permuted arm")
})

test_that("cap_at_target enforces alpha(n) <= target_fpr", {
  d <- sim_cell_type(40, n_genes = 40, seed = 9)
  out <- run_downsampling(d$counts, d$group, "wilcoxon",
                           grid_args = list(n_min = 10, n_points = 3),
                           k = 2, k_permute = 3, seed = 14)
  ac <- calibrate_alpha(out, target_fpr = 0.05, cap_at_target = TRUE)
  expect_true(all(ac$alpha <= 0.05))
})

test_that("predict_alpha interpolates and holds endpoints constant", {
  ac <- data.frame(n = c(10, 50, 100), alpha = c(0.01, 0.05, 0.1))
  expect_equal(predict_alpha(ac, 50), 0.05)
  expect_equal(predict_alpha(ac, 5), 0.01)
  expect_equal(predict_alpha(ac, 200), 0.1)
})

test_that("apply_adaptive_alpha recomputes n_degs without rerunning DE", {
  d <- sim_cell_type(40, n_genes = 40, seed = 10)
  out <- run_downsampling(d$counts, d$group, "wilcoxon",
                           grid_args = list(n_min = 10, n_points = 3),
                           k = 2, k_permute = 3, seed = 15)
  ac <- calibrate_alpha(out)
  ds2 <- apply_adaptive_alpha(out, ac)
  expect_true("alpha_used" %in% names(ds2))
  expect_equal(nrow(ds2), nrow(out$downsampled))
})

File: python/tests/test_adaptive_alpha.py
import numpy as np
import pandas as pd
import pytest

from burden.adaptive_alpha import apply_adaptive_alpha, calibrate_alpha, predict_alpha
from burden.downsample import run_downsampling


def test_calibrate_alpha_requires_permuted_arm(sim):
    d = sim(n_genes=40, n_per_group=30, seed=1)
    res = run_downsampling(
        d["counts"], d["group"], "wilcoxon", grid_args={"n_points": 3}, k=2, k_permute=0, seed=1
    )
    with pytest.raises(ValueError):
        calibrate_alpha(res)


def test_calibrate_alpha_shape_and_bounds(sim):
    d = sim(n_genes=60, n_per_group=40, seed=2)
    res = run_downsampling(
        d["counts"], d["group"], "wilcoxon", grid_args={"n_points": 3}, k=2, k_permute=3, seed=1
    )
    curve = calibrate_alpha(res, target_fpr=0.05)
    assert set(curve.columns) == {"n", "alpha", "n_null_tests"}
    assert (curve["alpha"] >= 0).all()
    assert list(curve["n"]) == sorted(curve["n"])


def test_calibrate_alpha_cap_at_target(sim):
    d = sim(n_genes=60, n_per_group=40, seed=3)
    res = run_downsampling(
        d["counts"], d["group"], "wilcoxon", grid_args={"n_points": 3}, k=2, k_permute=3, seed=1
    )
    curve = calibrate_alpha(res, target_fpr=0.05, cap_at_target=True)
    assert (curve["alpha"] <= 0.05).all()


def test_predict_alpha_interpolates_and_clamps():
    curve = pd.DataFrame({"n": [20, 50, 100], "alpha": [0.01, 0.03, 0.05]})
    out = predict_alpha(curve, [20, 35, 100, 500, 1])
    assert out[0] == pytest.approx(0.01)
    assert out[1] == pytest.approx(0.02)
    assert out[3] == pytest.approx(0.05)  # clamped above range
    assert out[4] == pytest.approx(0.01)  # clamped below range


def test_apply_adaptive_alpha_adds_column(sim):
    d = sim(n_genes=60, n_per_group=40, seed=4)
    res = run_downsampling(
        d["counts"], d["group"], "wilcoxon", grid_args={"n_points": 3}, k=2, k_permute=3, seed=1
    )
    curve = calibrate_alpha(res)
    out = apply_adaptive_alpha(res, curve)
    assert "alpha_used" in out.columns
    assert len(out) == len(res["downsampled"])
